src/clients/c++/simple_shm_client.cc

```diff
@@ -0,0 +1,374 @@
+    if (shm_addr == MAP_FAILED) {
+      std::cerr << "error: unable to process address space";
+      exit(1);
+    }
+  }
+
+  // Create  Output0 and Output1 in Shared Memory
+  // get shared memory file descriptor (NOT a file)
+  for (size_t n = 0; n < 2; n++) {
+    shm_key = "/output" + std::to_string(n) + "_simple";
+    shm_fd = shm_open(shm_key.c_str(), O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
+    if (shm_fd == -1) {
+      std::cerr << "error: unable to get output shared memory descriptor";
+      exit(1);
+    }
+    // extend shared memory object as by default it's initialized with size 0
+    res = ftruncate(shm_fd, output_byte_size);
+    if (res == -1) {
+      std::cerr << "error: unable to get initialize size";
+      exit(1);
+    }
+    // map shared memory to process address space
+    shm_addr = mmap(NULL, output_byte_size, PROT_READ, MAP_SHARED, shm_fd, 0);
+    if (shm_addr == MAP_FAILED) {
+      std::cerr << "error: unable to process address space";
+      exit(1);
+    }
+  }
+
+  // Write Input0 and Input1 to shared memory
+  for (size_t n = 0; n < 2; n++) {
+    shm_key = "/input" + std::to_string(n) + "_simple";
+    shm_fd = shm_open(shm_key.c_str(), O_RDWR, S_IRUSR | S_IWUSR);
+    if (shm_fd == -1) {
+      std::cerr << "error: unable to get shared memory descriptor";
+      exit(1);
+    }
+    shm_addr = mmap(NULL, input_byte_size, PROT_WRITE, MAP_SHARED, shm_fd, 0);
+    if (shm_addr == MAP_FAILED) {
+      std::cerr << "error: unable to process address space";
+      exit(1);
+    }
+    if (n == 0)
+      memcpy(shm_addr, &input0_data[0], input_byte_size);
+    else
+      memcpy(shm_addr, &input1_data[0], input_byte_size);
+  }
+
+  err = input0->SetSharedMemory("/input0_simple", 0, input_byte_size);
+  if (!err.IsOk()) {
+    std::cerr << "failed setting shared memory input: " << err << std::endl;
+    exit(1);
+  }
+  err = input1->SetSharedMemory("/input1_simple", 0, input_byte_size);
+  if (!err.IsOk()) {
+    std::cerr << "failed setting shared memory input: " << err << std::endl;
+    exit(1);
+  }
+
+  err = output0->SetSharedMemory("/output0_simple", 0, output_byte_size);
+  if (!err.IsOk()) {
+    std::cerr << "failed setting shared memory output: " << err << std::endl;
+    exit(1);
+  }
+  err = output1->SetSharedMemory("/output1_simple", 0, output_byte_size);
+  if (!err.IsOk()) {
+    std::cerr << "failed setting shared memory output: " << err << std::endl;
+    exit(1);
+  }
+
+
+  // Send inference request to the inference server.
+  std::map<std::string, std::unique_ptr<nic::InferContext::Result>> results;
+  FAIL_IF_ERR(infer_ctx->Run(&results), "unable to run model");
+
+  // We expect there to be 2 results and them to be written by TRTIS into the
+  // corresponding shared memory location provided. The 16 result elements
+  // in both outputs are the sum and difference calculated by the model.
+  for (size_t n = 0; n < 2; n++) {
```

Comment:
Don't open and memmap the region again...

Reply:
Right. I can re-used the `shm_addr` from the previous mmap